src/schema/sdl/finalizers.rs

```diff
@@ -0,0 +1,142 @@
+) -> &'a TypeRef<'a> {
+  match wrapper {
+      TypeWrapper::NonNull(inner) => {
+          arena.alloc(TypeRef::NonNullType(map_output_type(arena, inner)))
+      }
+
+      TypeWrapper::List(inner) => {
+          arena.alloc(TypeRef::ListType(map_output_type(arena, inner)))
+      }
+
+      TypeWrapper::Named(name) => {
+          arena.alloc(TypeRef::Type(name))
+      }
+  }
+}
+
+/// Converts a raw TypeDefinition into a schema type, but doesn't initialize everything, like object fields.
+/// Instead, it just creates the top level shells in order to allow referencing them in a second step (see finalizers).
+pub(super) fn initialize_type_definition<'a>(
+    ctx: &'a ASTContext,
+    typ: &'a TypeDefinition<'a>,
+) -> &'a SchemaType<'a> {
+    match typ {
+        TypeDefinition::ObjectTypeDefinition(obj) => {
+            let mut fields = HashMap::new_in(&ctx.arena);
+            for field in obj.fields.fields.iter() {
+              let schema_field = convert_to_schema_field(ctx, field.0, field.1);
+              fields.insert(*field.0, ctx.alloc(schema_field));
+            }
+
+            let mut interfaces = Vec::new_in(&ctx.arena);
+            for interface in obj.interfaces.iter() {
+                interfaces.push(*interface)
+            }
+
+            ctx.arena.alloc(SchemaType::Object(ctx.arena.alloc(SchemaObject {
+                name: obj.name,
+                fields,
+                interfaces
+            })))
+        }
+        TypeDefinition::InputObjectTypeDefinition(obj) => {
+            let mut fields = HashMap::new_in(&ctx.arena);
+            for (name, input_field) in obj.fields.fields.iter() {
+                fields.insert(
+                    *name,
+                    SchemaInputField {
+                        name,
+                        input_type: map_input_type(ctx, input_field.input_type),
+                    },
+                );
+            }
+
+            ctx.arena.alloc(SchemaType::InputObject(ctx.arena.alloc(SchemaInputObject {
+                name: obj.name,
+                fields
+            })))
+        }
+        TypeDefinition::EnumTypeDefinition(e) => ctx.arena.alloc(SchemaType::Enum(e)),
+        TypeDefinition::ScalarTypeDefinition(s) => ctx.arena.alloc(SchemaType::Scalar(s)),
+
+        TypeDefinition::InterfaceTypeDefinition(i) => {
+            let mut schema_interface = SchemaInterface::new(ctx, i.name);
+            for field in i.fields.fields.iter() {
+                let schema_field = convert_to_schema_field(ctx, field.0, field.1);
+                schema_interface.add_field(ctx, schema_field);
+            }
+
+            for obj in i.interfaces.iter() {
+                schema_interface.add_interface(ctx, obj);
+            }
+
+            for obj in i.possible_types.iter() {
+                schema_interface.add_possible_type(ctx, obj);
+            }
+
+            // TODO: possible interfaces aren't conveyed in the introspection
```

Comment:
Should there be post-processing to determine those? I' not sure what other SDL parsers do.

Reply:
Possible interfaces actually doesn't exist in the reference implementation, not sure why we added that